xitorch/optimize/rootfinder.py

```diff
@@ -1,393 +1,410 @@
+    # using the optimization algorithm, not the rootfinder algorithm
+    opt_method = method not in _RF_METHODS.keys()
+
+    # the rootfinder algorithms are designed to move to the opposite direction
+    # of the output of the function, so the output of this function is just
+    # the grad of z w.r.t. y
+    # if it is going to optimization method, then also returns the value
+
+    @make_sibling(pfunc)
+    def _min_fwd_fcn(y, *params):
+        with torch.enable_grad():
+            y1 = y.clone().requires_grad_()
+            z = pfunc(y1, *params)
+        grady, = torch.autograd.grad(z, (y1,), retain_graph=True,
+                                     create_graph=torch.is_grad_enabled())
+        return z, grady
+
+    @make_sibling(_min_fwd_fcn)
+    def _rf_fcn(y, *params):
+        z, grady = _min_fwd_fcn(y, *params)
+        return grady
+
+    # if using the optimization algorithm, then the forward function is the one
+    # that returns f and grad
+    if opt_method:
+        _fwd_fcn = _min_fwd_fcn
+    # if it is just using the rootfinder algorithm, then the forward function
+    # is the one that returns only the grad
+    else:
+        _fwd_fcn = _rf_fcn
+
+    alg_type = "minimizer" if opt_method else "rootfinder"
+    return _RootFinder.apply(_rf_fcn, y0, _fwd_fcn, alg_type, fwd_options, bck_options,
+                             len(params), *params, *pfunc.objparams())
+
+class _RootFinder(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, fcn, y0, fwd_fcn, alg_type, options, bck_options, nparams, *allparams):
+        # fcn: a function that returns what has to be 0 (will be used in the
+        #      backward, not used in the forward). For minimization, it is
+        #      the gradient
+        # fwd_fcn: a function that will be executed in the forward method
+        #          (unused in the backward)
+        # This class is also used for minimization, where fcn and fwd_fcn might
+        # be slightly different
+
+        # set default options
+        config = options
+        ctx.bck_options = bck_options
+        return_history = config.get("return_history", False)
+
+        params = allparams[:nparams]
+        objparams = allparams[nparams:]
+
+        with fwd_fcn.useobjparams(objparams):
+
+            method = config.pop("method")
+            methods = {
+                "minimizer": _OPT_METHODS,
+                "rootfinder": _RF_METHODS,
+                "equilibrium": _EQUIL_METHODS
+            }[alg_type]
+            name = alg_type
+            method_fcn = get_method(name, methods, method)
+            result = method_fcn(fwd_fcn, y0, params, **config)
+
+        ctx.fcn = fcn
+
+        # split tensors and non-tensors params
+        ctx.nparams = nparams
+        ctx.param_sep = TensorNonTensorSeparator(allparams)
+        tensor_params = ctx.param_sep.get_tensor_params()
+        if isinstance(result, tuple):
+            y, density_history = result
```

Comment:
Once again, please do not use the term density_history at xitorch level. You can call it rootfinder_history maybe.